sdks/java/core/src/main/java/org/apache/beam/sdk/util/SerializableUtils.java

```diff
@@ -79,16 +83,31 @@ public static Object deserializeFromByteArray(byte[] encodedValue,
   }
 
   public static <T extends Serializable> T ensureSerializable(T value) {
-    @SuppressWarnings("unchecked")
-    T copy = (T) deserializeFromByteArray(serializeToByteArray(value),
-        value.toString());
-    return copy;
+    return clone(value);
   }
 
   public static <T extends Serializable> T clone(T value) {
+    final Thread thread = Thread.currentThread();
+    final ClassLoader tccl = ReflectHelpers.findClassLoader();
+    ClassLoader loader = tccl;
+    try {
+      if (tccl.loadClass(value.getClass().getName()) != value.getClass()) {
+        loader = value.getClass().getClassLoader();
+      }
+    } catch (final NoClassDefFoundError | ClassNotFoundException e) {
+      loader = value.getClass().getClassLoader();
+    }
+    if (loader == null) {
+      loader = tccl; // will likely fail but the best we can do
+    }
+    thread.setContextClassLoader(loader);
```

Comment:
We can do this without overwriting the thread context class loader by passing forward which classloader we want to use into a new `deserializeFromByteArray` implementation.

The reason I suggest not setting the TCCL is that on line 109 your not restoring the original TCCL, your just setting it blindly to whatever ReflectHelpers detected (which isn't always the TCCL).

Reply:
@lukecwik I'm fine to revert to the original code which was handling TCCL correctly and not using ReflectHelper, is it ok?

Reply:
Yeah, I think it makes sense to drop usage of ReflectHelpers.

Reply:
Done with https://github.com/apache/beam/pull/3871/commits/c3837aa4ced89f63b1cbc376c01e4b52e361c767